mvrx/src/main/kotlin/com/airbnb/mvrx/MvCorStateStore.kt

```diff
@@ -0,0 +1,117 @@
+package com.airbnb.mvrx
+
+import io.reactivex.Observable
+import io.reactivex.disposables.Disposable
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.channels.actor
+import kotlinx.coroutines.channels.consumeEach
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.rx2.asObservable
+import java.util.*
+import kotlin.coroutines.CoroutineContext
+
+
+/**
+ * This is a container class around the actual state itself. It has a few optimizations to ensure
+ * safe usage of state.
+ *
+ * All state reducers are run in an actor to ensure that they don't have race
+ * conditions with each other.
+ *
+ */
+
+internal open class MvCorStateStore<S : Any>(initialState: S, final override val coroutineContext: CoroutineContext = Dispatchers.Default) : Disposable, CoroutineScope, MvRxStateStore<S> {
+
+    /**
+     * The stateChannel is where state changes should be pushed to.
+     */
+    private val stateChannel = ConflatedBroadcastChannel(initialState)
+    /**
+     * The observable observes the stateChannel but only emits events when the state actually changed.
+     */
+
+    override val observable: Observable<S> = stateChannel.openSubscription().asObservable(coroutineContext).distinctUntilChanged()
```

Comment:
if flipping to a coroutine store, would it make sense to go all the way and change `observable` to `flow`? cc @gpeal & @BenSchwab